[android] Simplify float toolbars and panels lifecycles

arnaudvergnet opened the pull request:
This PR tries to make showing and closing floating panels/toolbars easier. It encapsulates show and close functions in `MwmActivity` for easier management. It essentially adds functions to close all floating panels, called before showing a new one to ensure only one is shown at the same time.

I did not find any bugs from the simple testing I did, but it might require some more with different scenarios (especially on tablet).

Fixes #2005

biodranik commented on android/src/com/mapswithme/maps/MwmActivity.java:
```diff
@@ -663,12 +673,23 @@ private void closeAllFloatingPanels()
   }
 
-  public void closeMenu(@Nullable Runnable procAfterClose)
+  public boolean closeMenu()
```
You don't use this return value, is it really needed?

arnaudvergnet replied:
I use `closemenu()` on line 1092 in `onBackPressed` callback https://github.com/organicmaps/organicmaps/pull/2006/files#diff-957993735e10f75c9ccaaf6a794c90089c6e1e217682ea9aedcab85bfd806e36R1092. I don't use `closemenu(@Nullable Runnable procAfterClose)` return value but I added it for consistency.

arnaudvergnet replied:
The check is useful to know if calling the function actually closed the menu. This is useful in the `onBackPressed` function to stop when an action has been taken (to make sure a single press does not close everything at once). This is what was done in similar functions so I did the same for consistency.

But as `closemenu(@Nullable Runnable procAfterClose)` is never called in `onBackPressed`, the return value is not useful right now.

biodranik replied:
Let's don't copy old bad practices. If the return value is not needed then remove it. If it is used, then it makes sense at least to add a comment to this function (javadoc style) `// returns false if the menu was already closed.`

arnaudvergnet replied:
Fixed in https://github.com/organicmaps/organicmaps/pull/2006/commits/d273ffea50f3f61a3defb0eefe9c5c252c237e08

biodranik reviewed (commented):
Wow, that's an important and dangerous fix that should be carefully tested and reviewed.
@pastk @rtsisyk Please also help with testing.

There are many "modes" or dialogs in the OM now, so ideally everyone should be tested, for example, with the external API or ge0 link.

Did you also test an API click if the routing mode or navigation mode is active?

Did you test the tablet layout on Android Emulator?

arnaudvergnet commented:
> Wow, that's an important and dangerous fix that should be carefully tested and reviewed.

I don't know the app and all its dialogs as much as you do so your help will be needed!

> There are many "modes" or dialogs in the OM now

This is exactly why I thought a bit of refactoring was needed. I noticed a few bugs with floating panels so hopefully this PR will fix those and prevent new ones from occurring.

> Did you also test an API click if the routing mode or navigation mode is active?

I did not, but will do today.

> Did you test the tablet layout on Android Emulator?

I'll try to free some disk space to install the Tablet Emulator :<

arnaudvergnet commented:
I fixed the errors I found with navigation and planning. API urls work with and without navigation, and searching while planning or navigating works as well.

The function to close all floating panels and toolbars now has a new parameter to hide the search toolbar without stopping the search engine (useful when starting the planning state).

Please tell me if you find cases where this PR introduces regressions (I still have to try tablet mode).

biodranik commented:
Cases out of my head:
1. Search, Download, About, Bookmarks, Settings dialogs are already open before an API call.
2. "Search on map" is active.
3. Routing or navigation is active
4. User tried to build a route between different regions, and dialog offering to download these missing regions was active (or regions were in downloading progress with that dialog).
5. Some object was already selected on the map and it's Place Page information was open
6. Edit place dialog was open
7. Add a place mode was active
8. Top layers button was pressed
9. Bottom right menu was pressed
10. Previous API call already highlighted some points on the map

rtsisyk commented on android/src/com/mapswithme/maps/MwmActivity.java:
```diff
@@ -383,8 +383,7 @@ public void showDownloader(boolean openDownloaded)
     args.putBoolean(DownloaderActivity.EXTRA_OPEN_DOWNLOADED, openDownloaded);
```
@arnaudvergnet, thanks for contribution!
Is this patch a pure refactoring? Does it change UX? Please provide a video if UX is also changed. 
I've installed a beta version with your patch to one of my devices for testing. Let's see.

arnaudvergnet replied:
By refactoring, I also fixed some small issues like toolbars appearing over one another, and Android back button not stopping search while in navigation. So it changed UX but only to fix bugs.

I still have more testing to do based on the list provided by @biodranik, I'll try to keep UX changes to a minimum and will provide a video of what changed when testing is done.

rtsisyk replied:
Thanks a lot for explanation!

arnaudvergnet commented:
> Cases out of my head:
> 
>     1. Search, Download, About, Bookmarks, Settings dialogs are already open before an API call.
> 
>     2. "Search on map" is active.
> 
>     3. Routing or navigation is active
> 
>     4. User tried to build a route between different regions, and dialog offering to download these missing regions was active (or regions were in downloading progress with that dialog).
> 
>     5. Some object was already selected on the map and it's Place Page information was open
> 
>     6. Edit place dialog was open
> 
>     7. Add a place mode was active
> 
>     8. Top layers button was pressed
> 
>     9. Bottom right menu was pressed
> 
>     10. Previous API call already highlighted some points on the map

- 1 to 7: Everything works as expected from my testing. This PR should not have any impact on activities such as the editor and the downloader as it only changes toolbars and panels in the MwmActivity.
- 8: This menu behavior was not changed so it stays opened no matter what. I will include it in the function used to close all menu/panels/toolbars.
- 9 to 10: Everything works as expected. 

One issue I found so far is when you start a search and click on "view on map", then cancel and start a route (go into planning mode). The search wheel will show a blue icon as if you had search active, but clicking on it will start the search activity because no search is currently active. This may have to do with my latest commit so I will fix this.

Regarding UX changes, panels and toolbars should not be able to go over each other anymore. When in planning or navigation, if the user has the search wheel active (with a blue icon), pressing the android back button will cancel the search instead of closing the navigation. When in planning or navigation, if the user opens a place panel and resumes the activity (changing activity or switching and opening back the app), the panel will be closed. The last behavior is debatable so tell me if you want me to change it.

I will upload videos when I fix the last bugs.

**Note:** While testing, I had a strange issue with the API: When clicking a search link like om://search?cll=47.3813,8.5889&query=restaurants, it moves the camera over Zurich. I was able to reproduce it on the version from F-Droid as well so it has nothing to do with this PR. I can create an issue if you want.

biodranik commented:
> When in planning or navigation, if the user opens a place panel and resumes the activity (changing activity or switching and opening back the app), the panel will be closed.

Why do you want to close it? For example, I was building a route by selecting intermediate points. Then I got a call. Returned back to the app. WHERE IS MY SELECTED POINT? Ups…

arnaudvergnet commented:
> Why do you want to close it? For example, I was building a route by selecting intermediate points. Then I got a call. Returned back to the app. WHERE IS MY SELECTED POINT? Ups…

Did not think about this use-case. I created this behavior while refactoring and I tought it was quite nice, but yeah you are right I will revert it.

arnaudvergnet commented:
> When in planning or navigation, if the user opens a place panel and resumes the activity (changing activity or switching and opening back the app), the panel will be closed.

Fixed. It does not close anything anymore.

> One issue I found so far is when you start a search and click on "view on map", then cancel and start a route (go into planning mode). The search wheel will show a blue icon as if you had search active, but clicking on it will start the search activity because no search is currently active. This may have to do with my latest commit so I will fix this.

Fixed. I found this bug was present in versions before this PR, but it was simple to fix.

> While testing, I had a strange issue with the API: When clicking a search link like om://search?cll=47.3813,8.5889&query=restaurants, it moves the camera over Zurich. I was able to reproduce it on the version from F-Droid as well so it has nothing to do with this PR. I can create an issue if you want

This was not a bug, I just did not see the coordinates in the link.

**Here are some videos showing the new behaviors:**

- Mixing search, add place and api calls

https://user-images.githubusercontent.com/80701113/152646749-6400bcc9-33a6-462e-b3b8-d671eeef7de2.mp4

- Mixing search and navigation

https://user-images.githubusercontent.com/80701113/152646755-c7e42ab7-8c39-428b-b89a-db0fa63d702f.mp4

biodranik commented on android/src/com/mapswithme/maps/MwmActivity.java:
```diff
@@ -678,10 +704,60 @@ private boolean closePositionChooser()
+    return false;
+  }
 
+  private boolean closeSearchToolbar(boolean clearText, boolean stopSearch)
```
I don't see any case where stopSearch is false. Is this parameter really needed?

arnaudvergnet replied:
In `refreshSearchToolbar()` line 546, I call `closeFloatingToolbars(false, false)` which in turn calls this function with stopSearch to false. `refreshSearchToolbar()` is called on activity resume lifecycle event, so this allows hiding the search toolbar without stopping the search in case the app is in navigation or planning state (to keep the search wheel working). In case the app is not in navigation or planning, then the search toolbar is shown again.

biodranik replied:
Sorry, by some reason my browser did not show that diff.

biodranik commented on android/src/com/mapswithme/maps/routing/RoutingController.java:
```diff
@@ -648,20 +638,63 @@ public boolean cancel()
+
+  public void startPlanning(final @Nullable MapObject startPoint, final @Nullable MapObject endPoint)
+  {
+    if (mContainer != null)
```
Why is this check really needed in all these functions? When this variable can be null?

I'm asking because I know that in the past often developers did not investigate (and fix) the true cause, but did a "hotfix-workaround" instead.

arnaudvergnet replied:
I kept those checks because they were present before and I did not want to introduce new bugs/crashes.

From what I could see, `mContainer` is set in the `attach` method, called in the `MwmActivity` `onStart` lifecycle event. and deleted in the `onStop`.

As this field can be null, it is annotated with `@Nullable` thus Android Studio will complain when omitting this check.

biodranik replied:
Does it mean that these functions may be called when the main activity is stopped? If yes, what are these cases?

arnaudvergnet replied:
I don't think those functions can be called when the main activity is stopped, but as the field is nullable not having the check creates warnings.

Properly removing those checks without creating warnings would mean a bigger refactoring, probably to be done in an other PR.

arnaudvergnet commented:
By the way I see very few functions have documentation. What is the rule for this, should I write JavaDoc style comments for functions I add?

biodranik commented:
> By the way I see very few functions have documentation. What is the rule for this, should I write JavaDoc style comments for functions I add?

If the function's purpose is not clear from its name, then please add some docs. A clean and clear code doesn't need comments often.

arnaudvergnet commented:
I tried to clean up and use the new functions as much as possible to make future debugging easier. I also tried to test in tablet mode but I was not able to try navigation mode as for some reason the emulator for tablet does not want to find the GPS position.

Also the variable `mIsTabletLayout` is always set to false when I use the tablet emulator. I saw it was defined in `/res/values/attrs.xml`, should I manually set it to true here or is it supposed to automatically detect the tablet?

rtsisyk reviewed (approved):
I haven't found any problems during testing.
